docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,607 @@
+# ADR 038: KVStore state listening
+
+## Changelog
+
+- 11/23/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.
+
+## Context
+
+Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries)
+which proceed either through Tendermint and the ABCI, or through the gRPC server.
+In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.
+
+## Decision
+
+We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
+We will also introduce the tooling for writing these state changes out to files and configuring this service.
+
+### Listening interface
+
+In a new file, `store/types/listening.go`, we will create a `WriteListener` interface for streaming out state changes from a KVStore.
+
+```go
+// WriteListener interface for streaming data out from a listenkv.Store
+type WriteListener interface {
+	// if value is nil then it was deleted
```

Comment:
It would be good to separate writing `[]byte{}` and deleting.
It is technically possible to write an empty value, and it does remain in the iavl store (eg. via iterators)

Reply:
Hmm... I'm not sure this is true - I've tried that and gotten an error.